### PR TITLE
Add support for OTel Java runtime metrics

### PR DESCRIPTION
**Pull request by liustanley**
### What does this PR do?

- Moves runtime metric mappings into separate file
- Adds mappings for Java runtime metrics
- Removes `metricType` field from `runtimeMetricMapping`, changes code to use incoming metric type instead
- Refactors runtime metric attributes to support metrics that map from multiple attributes
- Changes mapping code to remove attributes that were used in the mapping process (i.e. when `jvm.heap_memory` is mapped, it doesn't copy the "type=heap" attribute from the OTel metric)
- Support mapping histogram runtime metrics

### CODE REVIEW

File: pkg/otlp/metrics/metrics_translator.go [hunk 1]
```diff
@@ -63,38 +62,69 @@ var runtimeMetricsMappings = map[string][]runtimeMetricMapping{
...
 		attribute:      "generation",
 		attributeValue: "gen1",
-		metricType:     pmetric.MetricTypeSum,
 	}, {
 		mappedName:     "runtime.dotnet.gc.count.gen2",
 		attribute:      "generation",
 		attributeValue: "gen2",
-		metricType:     pmetric.MetricTypeSum,
+	}},
+	"process.runtime.jvm.threads.count": {{mappedName: "jvm.thread_count"}},
```
[the remark this answers is not in the document]
Comment (liustanley) [reply pkg/otlp/metrics/metrics_translator.go:91]: Sure, should we move the mappings to a new file since they're getting kind of large?

**Review by dineshg13 — CHANGES_REQUESTED** [review]
Can we also add GC metrics ? 
image


https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6964/files#diff-bb4826b5d07256ae1befe69927b73586a37da8bb6dcdfa30a3ba18e5c828b55b

File: pkg/otlp/metrics/runtime_metric_mappings.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+// runtimeMetricAttribute defines the structure for an attribute in regard to mapping runtime metrics.
+// The presence of a runtimeMetricAttribute means that a metric must be mapped from a data point
+// with the given attribute(s).
+type runtimeMetricAttribute struct {
+	key    string   // the attribute name
+	values []string // the attribute value, or multiple values if there is more than one value for the same mapping
+}
+
+var goRuntimeMetricsMappings = map[string][]runtimeMetricMapping{
```
Comment (dineshg13) [inline pkg/otlp/metrics/runtime_metric_mappings.go:18]: Every runtime mapping should follow this silent contract - `map[string][]runtimeMetricMapping` --- Can we create this as a try or interface ?

**Review by dineshg13 — APPROVED** [review]
Looks good. Added a small comment

**Comment by sirianni** [comment]
The aliasing of `process.runtime.jvm.gc.duration` to `jvm.gc.parnew.time` breaks setups that have both legacy Java apps using JMX and new Java applications using OTel due to a incompatibility in metric type (`distribution` vs. `gauge`).

In these scenarios, the `jvm.gc.parnew.time` metric metadata on the Datadog backend gets changed from a `gauge` to a `distribution` which causes the JMX-fetched gauge variant of `jvm.gc.parnew.time` to be completely unqueryable (disappears from all dashboards and monitors).

**Comment by mx-psi** [comment]
@sirianni can you file a support ticket for this? Please share the case number here and I can loop in the relevant people :)

**Comment by sirianni** [comment]
> Please share the case number here and I can loop in the relevant people :)

https://help.datadoghq.com/hc/en-us/requests/1434627

**Pull request merged**